pkg/log/text_formatter.go

```diff
@@ -15,191 +15,67 @@
 package log
 
 import (
+	"bufio"
 	"bytes"
-	"fmt"
-	"runtime"
-	"sort"
 	"strings"
-	"time"
 
 	"github.com/Sirupsen/logrus"
 )
 
-const (
-	nocolor = 0
-	red     = 31
-	green   = 32
-	yellow  = 33
-	blue    = 34
-	gray    = 37
-)
-
-var (
-	baseTimestamp time.Time
-	isTerminal    bool
-)
-
-func init() {
-	baseTimestamp = time.Now()
-	isTerminal = logrus.IsTerminal()
-}
-
-func miniTS() int {
-	return int(time.Since(baseTimestamp) / time.Second)
-}
-
 type TextFormatter struct {
-
-	// Enable colors
-	Colors bool
-
-	// Enable timestamp logging. It's useful to disable timestamps when output
-	// is redirected to logging system that already adds timestamps.  When
-	// disabled, a delta is used for each log line from the start of the
-	// process.  Enabling will apply a timestamp derived from TimestampFormat.
-	Timestamp bool
-
 	// TimestampFormat is the format used to print the timestamp.  By default
 	// an RFC3339 timestamp is used.
 	TimestampFormat string
-
-	// The fields are sorted by default for a consistent output. For applications
-	// that log extremely frequently and don't use the JSON formatter this may not
-	// be desired.
-	DisableSorting bool
-
-	// Escape noncharacter ascii strings.
-	EscapeNonCharacters bool
 }
 
-// NewTextFormatter returns a text formatter with defaults appropriate for the
-// TTY or file being written to.
+// NewTextFormatter returns a text formatter
 func NewTextFormatter() *TextFormatter {
+	return &TextFormatter{
+		TimestampFormat: "Jan _2 2006 15:04:05.000Z07:00",
+	}
+}
 
-	isColorTerminal := isTerminal && (runtime.GOOS != "windows")
-
-	formatter := &TextFormatter{
-		Colors:          isColorTerminal,
-		Timestamp:       false,
-		TimestampFormat: time.RFC3339,
-		DisableSorting:  false,
+func trimOrPadToSize(s string, size int) string {
+	if len(s) > size {
+		return s[:size]
 	}
 
-	return formatter
+	return s + strings.Repeat(" ", size-len(s))
 }
 
 func (f *TextFormatter) Format(entry *logrus.Entry) ([]byte, error) {
-	var b *bytes.Buffer
-	keys := make([]string, 0, len(entry.Data))
-	for k := range entry.Data {
-		keys = append(keys, k)
+	t := f.timeStamp(entry)
+	l := strings.ToUpper(trimOrPadToSize(entry.Level.String(), 5))
+
+	var b bytes.Buffer
+
+	if entry.Message == "" {
+		b.Grow(len(t) + len(" ") + len(l) + len("\n"))
```

Comment:
Trying to avoid reallocating buffer multiple times. I tried several combinations; this ones seems to result in the least amount of allocations.